File: packages/go/internal/vdata/vdata.go
// Package vdata caches a vector's materialized datasets so multi-megabyte
// $prng streams are generated once per vector, however many times their bytes
// or derived values (${data.<name>.<field>}) are referenced.
package vdata

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"hash/crc32"
	"hash/crc64"
	"strconv"

	s3vectors "github.com/cloud-portable/s3vectors/packages/go"
	"github.com/cloud-portable/s3vectors/packages/go/datagen"
)

var (
	crc32cTable = crc32.MakeTable(crc32.Castagnoli)
	crc64Table  = crc64.MakeTable(0x9A6C9329AC4BC9B5) // CRC-64/NVME, as in datagen
)

// Cache materializes and memoizes one vector's datasets. Not safe for
// concurrent use; each vector executes on a single goroutine.
type Cache struct {
	specs map[string]s3vectors.DataSpec
	bytes map[string][]byte
}

func New(specs map[string]s3vectors.DataSpec) *Cache {
	return &Cache{specs: specs, bytes: map[string][]byte{}}
}

// StreamThreshold is the dataset size above which a caller should stream rather
// than materialize. Below it, datasets are small enough that caching wins:
// bodies and digests reference the same dataset repeatedly, and one
// materialization plus N local hashes beats N re-reads. Above it, holding the
// bytes is what makes a gigabyte-scale vector unrunnable.
const StreamThreshold = 8 << 20

// Size reports the dataset's declared length in bytes, generating nothing.
func (c *Cache) Size(name string) (int64, error) {
	return datagen.Size(c.specs, name)
}

// Reader returns a seekable reader over the dataset. It generates on demand and
// is deliberately not cached — the point is to never hold the bytes. The reader
// is an io.Seeker, which the AWS SDK needs to derive Content-Length and to
// rewind after hashing the payload.
func (c *Cache) Reader(name string) (*datagen.Reader, error) {
	return datagen.NewReader(c.specs, name)
}

// Bytes returns the dataset's bytes (generated on first use). The returned
// slice is shared — callers must not mutate it.
func (c *Cache) Bytes(name string) ([]byte, error) {
	if b, ok := c.bytes[name]; ok {
		return b, nil
	}
	b, err := datagen.Generate(c.specs, name)
	if err != nil {
		return nil, err
	}
	c.bytes[name] = b
	return b, nil
}

// Derived computes a ${data.<name>.<field>} value from the cached bytes.
// Field semantics mirror datagen.Derived exactly.
func (c *Cache) Derived(name, field string) (string, error) {
	// Above the threshold, delegate to the corpus: its Derived digests in
	// chunks, where this cache would have to hold the whole dataset.
	if n, err := c.Size(name); err == nil && n > StreamThreshold {
		return datagen.Derived(c.specs, name, field)
	}
	b, err := c.Bytes(name)
	if err != nil {
		return "", err
	}
	switch field {
	case "size":
		return strconv.Itoa(len(b)), nil
	case "md5":
		sum := md5.Sum(b)
		return hex.EncodeToString(sum[:]), nil
	case "etag":
		sum := md5.Sum(b)
		return `"` + hex.EncodeToString(sum[:]) + `"`, nil
	case "sha256":
		sum := sha256.Sum256(b)
		return hex.EncodeToString(sum[:]), nil
	case "sha256B64":
		sum := sha256.Sum256(b)
		return base64.StdEncoding.EncodeToString(sum[:]), nil
	case "sha1B64":
		sum := sha1.Sum(b)
		return base64.StdEncoding.EncodeToString(sum[:]), nil
	case "crc32B64":
		var v [4]byte
		binary.BigEndian.PutUint32(v[:], crc32.ChecksumIEEE(b))
		return base64.StdEncoding.EncodeToString(v[:]), nil
	case "crc32cB64":
		var v [4]byte
		binary.BigEndian.PutUint32(v[:], crc32.Checksum(b, crc32cTable))
		return base64.StdEncoding.EncodeToString(v[:]), nil
	case "crc64nvmeB64":
		var v [8]byte
		binary.BigEndian.PutUint64(v[:], crc64.Checksum(b, crc64Table))
		return base64.StdEncoding.EncodeToString(v[:]), nil
	default:
		return "", fmt.Errorf("unknown derived data field: %s", field)
	}
}

File: packages/python/src/cloud_portable_s3tests/_vdata.py
"""Per-vector dataset cache: multi-megabyte ``$prng`` streams are generated
once per vector, however many times their bytes or derived values
(``${data.<name>.<field>}``) are referenced. Derived fields are computed
locally from those cached bytes — the corpus ``derived`` is bounded in memory
but re-reads the dataset per field, so one materialization plus N local hashes
is cheaper here. The semantics mirror the datagen reference and are asserted
equal against it in the tests."""

from __future__ import annotations

import base64
import io
import hashlib
import struct
import zlib
from typing import Any, Mapping

from cloud_portable_s3vectors.datagen import data_size, generate, generate_range

#: Dataset size above which a caller should stream rather than materialize.
#: Below it, caching wins: bodies and digests reference the same dataset
#: repeatedly. Above it, holding the bytes is what makes a gigabyte-scale vector
#: unrunnable. Keep in step with the Go runner's ``vdata.StreamThreshold``.
STREAM_THRESHOLD = 8 * 1024 * 1024


class DatasetReader(io.RawIOBase):
    """A seekable read-only file object over a dataset, for boto3's ``Body``.

    The corpus exposes an iterator of chunks; botocore wants ``read(n)`` and
    seeks the body to size it and to rewind after signing, so reads are served
    as ranges instead. Nothing is cached: the point is to never hold the bytes.
    """

    def __init__(self, specs, name: str) -> None:
        self._specs = specs
        self._name = name
        self._size = data_size(specs, name)
        self._pos = 0

    def readable(self) -> bool:
        return True

    def seekable(self) -> bool:
        return True

    def __len__(self) -> int:
        return self._size

    def read(self, size: int = -1) -> bytes:
        if size is None or size < 0:
            size = self._size - self._pos
        size = min(size, self._size - self._pos)
        if size <= 0:
            return b""
        chunk = generate_range(self._specs, self._name, self._pos, size)
        self._pos += size
        return chunk

    def readinto(self, buf) -> int:
        chunk = self.read(len(buf))
        buf[: len(chunk)] = chunk
        return len(chunk)

    def seek(self, offset: int, whence: int = io.SEEK_SET) -> int:
        if whence == io.SEEK_SET:
            pos = offset
        elif whence == io.SEEK_CUR:
            pos = self._pos + offset
        elif whence == io.SEEK_END:
            pos = self._size + offset
        else:
            raise ValueError(f"invalid whence {whence}")
        if pos < 0:
            raise ValueError(f"negative position {pos}")
        self._pos = pos
        return pos

    def tell(self) -> int:
        return self._pos


class DataError(ValueError):
    """An unknown dataset, field or invalid slice."""


class DataCache:
    def __init__(self, specs: Mapping[str, Any] | None) -> None:
        self.specs: Mapping[str, Any] = specs if specs is not None else {}
        self._bytes: dict[str, bytes] = {}
        self._derived: dict[tuple[str, str], str] = {}

    def bytes(self, name: str) -> bytes:
        """Dataset bytes, generated on first use (shared; never mutate)."""
        b = self._bytes.get(name)
        if b is None:
            try:
                b = generate(self.specs, name)
            except (KeyError, ValueError) as err:
                raise DataError(str(err).strip("'\"")) from err
            self._bytes[name] = b
        return b

    def size(self, name: str) -> int:
        """The dataset's declared length in bytes, generating nothing."""
        return data_size(self.specs, name)

    def reader(self, name: str) -> DatasetReader:
        """A seekable file object over the dataset; never cached."""
        return DatasetReader(self.specs, name)

    def derived(self, name: str, field: str) -> str:
        """A ``${data.<name>.<field>}`` value, memoized and computed from the
        cached bytes."""
        key = (name, field)
        v = self._derived.get(key)
        if v is None:
            v = derive_field(self.bytes(name), field)
            self._derived[key] = v
        return v


def derive_field(data: bytes, field: str) -> str:
    if field == "size":
        return str(len(data))
    if field == "md5":
        return hashlib.md5(data).hexdigest()
    if field == "etag":
        return f'"{hashlib.md5(data).hexdigest()}"'
    if field == "sha256":
        return hashlib.sha256(data).hexdigest()
    if field == "sha256B64":
        return base64.b64encode(hashlib.sha256(data).digest()).decode("ascii")
    if field == "sha1B64":
        return base64.b64encode(hashlib.sha1(data).digest()).decode("ascii")
    if field == "crc32B64":
        return base64.b64encode(struct.pack(">I", zlib.crc32(data) & 0xFFFFFFFF)).decode("ascii")
    if field == "crc32cB64":
        return base64.b64encode(struct.pack(">I", _crc32c(data))).decode("ascii")
    if field == "crc64nvmeB64":
        return base64.b64encode(struct.pack(">Q", _crc64nvme(data))).decode("ascii")
    raise DataError(f"unknown derived data field: {field}")


# The checksum implementations mirror the corpus datagen reference; the vdata
# tests assert equality with the corpus package's own derived() for every field.


def _make_table(poly: int, width: int) -> list[int]:
    mask = (1 << width) - 1
    table = []
    for n in range(256):
        c = n
        for _ in range(8):
            c = ((c >> 1) ^ poly) if c & 1 else (c >> 1)
        table.append(c & mask)
    return table


_CRC32C_TABLE = _make_table(0x82F63B78, 32)
_CRC64_TABLE = _make_table(0x9A6C9329AC4BC9B5, 64)


def _crc32c(data: bytes) -> int:
    c = 0xFFFFFFFF
    t = _CRC32C_TABLE
    for b in data:
        c = t[(c ^ b) & 0xFF] ^ (c >> 8)
    return (c ^ 0xFFFFFFFF) & 0xFFFFFFFF


def _crc64nvme(data: bytes) -> int:
    c = 0xFFFFFFFFFFFFFFFF
    t = _CRC64_TABLE
    for b in data:
        c = t[(c ^ b) & 0xFF] ^ (c >> 8)
    return c ^ 0xFFFFFFFFFFFFFFFF
